### PR TITLE
Fix contiguity analysis of predicates to match updated contiguity.

### PR DESCRIPTION
**Pull request by csarofeen**
This PR depends on https://github.com/csarofeen/pytorch/pull/1990

Extends contiguity method introduced there to predicate analysis. A few questions marked in comments but should be safe as is so the questions could be pushed to a follow up.

```.cpp
  int w = 15, x = 31;

  auto tv0 = makeContigTensor(4);
  fusion.addInput(tv0);
  auto tv1 = sin(tv0);
  auto tv2 = view(tv1, {w, x}, {x, w});

  fusion.addOutput(tv2);

  tv2->merge(0)->split(0, 4)->split(0, 8, false);

  TransformPropagator propagator(tv2);
  MaxRootDomainInfoSpanningTree spanning_tree(tv2);
  spanning_tree.traverse(&propagator);
```
Will now generate:
```.cpp
__global__ void CUDAGeneratedKernel(Tensor<float, 2> T0, Tensor<float, 2> T2) {
  NVFUSER_DEFINE_MAGIC_ZERO
  float T1[((8 * (ceilDiv((ceilDiv((31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))), 4)), 8))) * 4)];
  #pragma unroll
  for(nvfuser_index_t i28 = 0; i28 < 8; ++i28) {
    #pragma unroll 1
    for(nvfuser_index_t i29 = 0; i29 < (ceilDiv((ceilDiv((31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))), 4)), 8)); ++i29) {
      #pragma unroll
      for(nvfuser_index_t i30 = 0; i30 < 4; ++i30) {
        int64_t i66;
        i66 = (((i28 * (ceilDiv((ceilDiv((31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))), 4)), 8))) + i29) * 4) + (i30 + nvfuser_zero);
        if ((i66 < (31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))))) {
          T1[((((i28 * (ceilDiv((ceilDiv((31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))), 4)), 8))) + i29) * 4) + i30)]
             = sinf(T0[i66]);
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  #pragma unroll
  for(nvfuser_index_t i31 = 0; i31 < 8; ++i31) {
    #pragma unroll 1
    for(nvfuser_index_t i32 = 0; i32 < (ceilDiv((ceilDiv((31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))), 4)), 8)); ++i32) {
      #pragma unroll
      for(nvfuser_index_t i33 = 0; i33 < 4; ++i33) {
        int64_t i151;
        i151 = (((i31 * (ceilDiv((ceilDiv((31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))), 4)), 8))) + i32) * 4) + (i33 + nvfuser_zero);
        if ((i151 < (31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))))) {
          T2[i151]
             = T1[((((i31 * (ceilDiv((ceilDiv((31 * (ceilDiv((T0.size[0] * T0.size[1]), 31))), 4)), 8))) + i32) * 4) + i33)];
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
}
```

No significant perf change detected with this PR which is expected
image

### CODE REVIEW

**Comment by naoyam** [comment]
This PR should close #2001

**Comment by csarofeen** [comment]
@naoyam this should be ready for review.

File: torch/csrc/jit/codegen/cuda/iter_visitor.h [hunk 1]
```diff
@@ -311,6 +311,29 @@ class StmtSort : public IterVisitor {
...
+  // i3 = i2
+  // i4 = i3 + i1
+  // getExprs(fusion, {i4}, {i3})
+  // will return the definition and values {i0, i1, i4}
+  // i3 is dependent on i1, but since i4 also is then the traversal will go down
+  // the i4->i1->i0 path, even though the i4->i3-//>i2->i1 path is blocked.
+  //
+  // If traverse_members it will also extract all member nodes in the sorted
+  // expr list in the fusion. i.e. all expressions on IterDomains, extents, etc
+  static std::vector<Statement*> getStmts(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/iter_visitor.h]: Is this a duplicate of `getStmtsBetween`? I don't see its implementation in iter_visitor.cpp. Is it just accidentally left here after renaming to `getStmtsBetween`?

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
+
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: Hahaha, lower_divisble_split.cpp never dies!

File: torch/csrc/jit/codegen/cuda/contiguity.cpp [hunk 3]
```diff
@@ -520,9 +528,19 @@ void ContigIDs::handle(Merge* merge) {
...
-      if (!root_contiguity_[root_id_i] && !root_ids.empty()) {
-        // Otherwise this merge->out() isn't a contiguously indexable ID
-        return;
+      if (!root_contiguity_[root_id_i]) {
+        // If it's the last root, and we're indexing
+        // (!ignore_consistent_ordering) we can still consider this ID
+        // contiguously indexable since it will still be multiplied by its
+        // stride. If we're computing predicates, then we don't want to do this,
+        // as when we mark something as non-contiguous we don't want it merged
+        // with any other domains.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/contiguity.cpp]: I don't follow this. First of all, are we always assuming tensors are fully contiguous when generating predicates? That's what I remember, and if it's still the case, this part of the code should never be reachable for predicates.

**Review by naoyam — COMMENTED** [review]
Comments so far

File: torch/csrc/jit/codegen/cuda/iter_visitor.h (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.h]: Shoot, this snuck through some how. Had conflicts when I merged and wanted to keep the comment, not the signature, will fix.

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: sorry, git is really failing me, tried to patch it up last night and guess I was careless yet again.

File: torch/csrc/jit/codegen/cuda/contiguity.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/contiguity.cpp]: I just tried to split up the conditional into two stages. The predicate path gets caught below in `!ignore_consistent_ordering_` sorry for the confusion. Predicate will have `ignore_consistent_ordering_ == true` so it won't exit/return non-contiguous based on this check.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/contiguity.cpp]: Comment is confusing, let me take another attempt at it.

File: torch/csrc/jit/codegen/cuda/iter_visitor.h (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.h]: Yeah, just a merge conflict, will remove.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 4]
```diff
@@ -2332,32 +2332,17 @@ std::vector<PredicateDomainInfo> getPredicateContigIds(
...
+  for (auto entry : consumer_index_map) {
+    auto c_id = gpu_lower->caMap()->getConcreteMappedID(
+        entry.first, IdMappingMode::EXACT);
+    contig_concrete_to_ref_map[c_id] = c_id;
+  }
+  std::vector<bool> predicate_contiguity(consumer_root_domain.size(), true);
+  for (auto root_i : c10::irange(predicate_contiguity.size())) {
+    auto root_id = consumer_root_domain[root_i];
+    if (root_id->maybePartial()) {
+      // predicate_contiguity[root_i] = false;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Is this safe to mark it as contiguous? Probably no unit test exercises this case, but as it was commented, predicates should be done at root domains in this case.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 5]
```diff
@@ -2366,63 +2351,50 @@ std::vector<PredicateDomainInfo> getPredicateContigIds(
...
-    if ((shift_expr && shift_expr->offset(consumer_root_pos) != 0) ||
-        (gather_expr && consumer_root_pos < gather_expr->windowShape().size() &&
-         gather_expr->windowShape().at(consumer_root_pos) != 1)) {
-      excluded_ids.insert(consumer_root_id);
+    if ((shift_expr && shift_expr->offset(root_i) != 0) ||
+        (gather_expr && root_i < gather_expr->windowShape().size() &&
+         gather_expr->windowShape().at(root_i) != 1)) {
+      // TODO: The following line commented out didn't have any failures, is it
+      // needed?
+      predicate_contiguity[root_i] = false;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: I think this is necessary. It probably just doesn't have tests. Let me add tests.

File: torch/csrc/jit/codegen/cuda/contiguity.cpp [hunk 6]
```diff
@@ -512,17 +520,27 @@ void ContigIDs::handle(Merge* merge) {
...
-      if (!root_contiguity_[root_id_i] && !root_ids.empty()) {
-        // Otherwise this merge->out() isn't a contiguously indexable ID
-        return;
+      // If we're indexing:
+      // we could still potentially consider this ID linearly indexable, as we
+      // could multiple the index by the last root's stride.
+      //
+      // If we're computing predicates (ignore_consistent_ordering_==true),
+      // then we don't have this same constraint, we can just ignore
+      // contiguity of the roots all together.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/contiguity.cpp:550]: Makes sense. Thanks for clarification.

File: torch/csrc/jit/codegen/cuda/contiguity.cpp [hunk 7]
```diff
@@ -512,17 +520,27 @@ void ContigIDs::handle(Merge* merge) {
...
+      if (!root_contiguity_[root_id_i] && is_indexing_pass) {
+        // TODO: This didn't error when I removed "!ignore_consistent_ordering_"
+        // is it really needed?
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/contiguity.cpp]: Do you still want to keep this comment? It should actually not be necessary. If this is used for indexing, `is_indexing_pass` is true, so nothing should change. If used for predicates, `!root_contiguity_` should be always true, so `is_indexing_pass` doesn't matter.

I actually like the current code as it's easier to understand the underlying logic.

File: torch/csrc/jit/codegen/cuda/contiguity.cpp (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/contiguity.cpp:550]: Actually, this isn't correct. Even for predicates, some root domains are marked as non-contiguous when they need to be predicated at root domains.

Some cleanups would be needed here. I'll work on it.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: Fixed and added a unit test (`FusionContigPredicateShift`). The test should fail without the fix.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: Fixed and added a unit test (FusionContigPredicateShift). The test should fail without the fix.

File: torch/csrc/jit/codegen/cuda/contiguity.cpp (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/contiguity.cpp:550]: A problem with the design is `root_contiguity` is used both as indicating contiguity AND if domains are not necessary to be predicated at root. These are two separate concepts, so I split out the latter as `final_ids`.

**Review by naoyam — APPROVED** [review]
Stamping for now

**Comment by naoyam** [comment]
@csarofeen I pushed a few commits. Please let me know if they make sense. I already approved the PR so that this can be merged if they look good.

**Comment by csarofeen** [comment]
LGTM Thanks!

**Pull request merged**